# Data Cleaning and Manipulation Activity
In this activity, you'll clean and prepare two datasets, **Product Inventory** and **Sales Data**, for analysis. 

## Task 1: Data Import
1. Import the datasets.
    - For the CSV file: Use read_csv. 
    - For the Excel file: Use read_excel.
2. Inspects the structure of the datasets.

In [10]:

# Import necessary libraries
import pandas as pd  # Pandas library is essential for data manipulation

# Task 1: Import the datasets
# TODO Import the sales data (../data/sales_data.csv)
sales_data = pd.read_csv("sales_data.csv") # Load the sales data CSV file

# TODO Import the product inventory data (../data/product_inventory.xlsx)
inventory_data = pd.read_excel("product_inventory.xlsx") # Load the inventory data Excel file

# Inspect the sales data
print("Sales Data Info:")
print(sales_data.info())  # Inspect structure and data types of sales data
print(sales_data.head())  # View the first few rows of sales data
# print(sales_data) # if you want to see the whole data set

# Inspect the inventory data
print("\nInventory Data Info:")
print(inventory_data.info())  # Inspect structure and data types of inventory data
print(inventory_data.head())  # View the first few rows of inventory data
# print(inventory_data) # if you want to see the whole data set

Sales Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   SaleID      5 non-null      int64  
 1   Product_ID  5 non-null      int64  
 2   Quantity    5 non-null      int64  
 3   Price       4 non-null      float64
 4   SaleDate    5 non-null      object 
dtypes: float64(1), int64(3), object(1)
memory usage: 328.0+ bytes
None
   SaleID  Product_ID  Quantity  Price    SaleDate
0       1         101         2  19.99  2024-11-01
1       2         102         1  29.99  2024-11-02
2       3         103         3    NaN  2024-11-03
3       4         101         5  19.99  2024-11-04
4       5         104         2  39.99  2024-11-05

Inventory Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   ProductID

## Task 2: Data Cleaning
1. Clean the data.

    - When cleaning data, look incosistent format issues and consider the appropriate methods to address them.
    - Focus on making the datasets as complete and compatible as possible before moving to the next step. 
    - Test your work by inspecting the datasets after cleaning.

**Checklist:**

-   Have you checked for and handled missing values?
-   Have you removed duplicate rows if necessary?
-   Are all columns in the correct format?
-   Are column names consistent across datasets?

In [ ]:

# Task 2: Data Cleaning
# TODO: Clean the data
    

SaleID        0
Product_ID    0
Quantity      0
Price         0
SaleDate      0
dtype: int64
SaleID                 int64
Product_ID             int64
Quantity               int64
Price                float64
SaleDate      datetime64[ns]
dtype: object
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   ProductID    5 non-null      int64 
 1   ProductName  5 non-null      object
 2   Stock        5 non-null      int64 
dtypes: int64(2), object(1)
memory usage: 248.0+ bytes
None

Cleaned Sales Data:
   SaleID  Product_ID  Quantity  Price   SaleDate
0       1         101         2  19.99 2024-11-01
1       2         102         1  29.99 2024-11-02
2       3         103         3  27.49 2024-11-03
3       4         101         5  19.99 2024-11-04
4       5         104         2  39.99 2024-11-05

Cleaned Inventory Data:
   ProductID    ProductName  Stock
0 

In [18]:
# Check of missing value
print(sales_data.isnull().sum())
print(inventory_data.isnull().sum())

SaleID        0
Product_ID    0
Quantity      0
Price         0
SaleDate      0
dtype: int64
ProductID      0
ProductName    0
Stock          0
dtype: int64


In [19]:
# Check data types of all columns
print(sales_data.dtypes)
print(inventory_data.info())

SaleID                 int64
Product_ID             int64
Quantity               int64
Price                float64
SaleDate      datetime64[ns]
dtype: object
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   ProductID    5 non-null      int64 
 1   ProductName  5 non-null      object
 2   Stock        5 non-null      int64 
dtypes: int64(2), object(1)
memory usage: 248.0+ bytes
None


In [22]:
#Fill missing value in Price with average value
sales_data['Price'] = sales_data['Price'].fillna(sales_data['Price'].mean())

In [23]:
#Convert datetime in correct format for SaleDate
sales_data['SaleDate'] = pd.to_datetime(sales_data['SaleDate'],errors='coerce')

In [24]:
# Verify cleaning steps
print("\nCleaned Sales Data:")
print(sales_data.head())
print("\nCleaned Inventory Data:")
print(inventory_data.head())


Cleaned Sales Data:
   SaleID  Product_ID  Quantity  Price   SaleDate
0       1         101         2  19.99 2024-11-01
1       2         102         1  29.99 2024-11-02
2       3         103         3  27.49 2024-11-03
3       4         101         5  19.99 2024-11-04
4       5         104         2  39.99 2024-11-05

Cleaned Inventory Data:
   ProductID    ProductName  Stock
0        101       Widget A     10
1        102       Gadget B      5
2        103  Thingamajig C      7
3        104    Doohickey D      8
4        105  Contraption E      3


In [25]:
# Standardise text columns by stripping spaces
inventory_data['ProductName'] = inventory_data['ProductName'].str.strip()

In [26]:
#View unique values after cleaning
print(inventory_data['ProductName'].unique())


['Widget A' 'Gadget B' 'Thingamajig C' 'Doohickey D' 'Contraption E']


In [28]:
# Rename mismatched column to match the current dataset
inventory_data.rename(columns={'ProductID': 'Product_ID'}, inplace=True)
print(inventory_data)


   Product_ID    ProductName  Stock
0         101       Widget A     10
1         102       Gadget B      5
2         103  Thingamajig C      7
3         104    Doohickey D      8
4         105  Contraption E      3


In [ ]:
#Final sales data for further analysis
sales_data.to_csv("final_sales_data.csv")

In [31]:
final_sales_data = pd.read_csv("final_sales_data.csv")
print(final_sales_data.head())

   Unnamed: 0  SaleID  Product_ID  Quantity  Price    SaleDate
0           0       1         101         2  19.99  2024-11-01
1           1       2         102         1  29.99  2024-11-02
2           2       3         103         3  27.49  2024-11-03
3           3       4         101         5  19.99  2024-11-04
4           4       5         104         2  39.99  2024-11-05


In [32]:
#Final inventory data for further analysis
inventory_data.to_csv("final_inventory_data.csv")

In [33]:
final_inventory_data = pd.read_csv("final_inventory_data.csv")
print(final_inventory_data.head())

   Unnamed: 0  Product_ID    ProductName  Stock
0           0         101       Widget A     10
1           1         102       Gadget B      5
2           2         103  Thingamajig C      7
3           3         104    Doohickey D      8
4           4         105  Contraption E      3


## Task 3: Data Merging
1. Merge the sales and inventory data.

In [42]:

# Task 3: Data Merging
# Merge sales and inventory data
# TODO: Merge the datasets. Replace ColumnName with the key column name
merged_data = pd.merge(final_sales_data, final_inventory_data, on='Product_ID', how='left') 
    

In [43]:
# Inspect merged data
print("\nMerged Data Info:")
print(merged_data.info())
print(merged_data.head())


Merged Data Info:
<class 'pandas.core.frame.DataFrame'>
Int64Index: 5 entries, 0 to 4
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0_x  5 non-null      int64  
 1   SaleID        5 non-null      int64  
 2   Product_ID    5 non-null      int64  
 3   Quantity      5 non-null      int64  
 4   Price         5 non-null      float64
 5   SaleDate      5 non-null      object 
 6   Unnamed: 0_y  5 non-null      int64  
 7   ProductName   5 non-null      object 
 8   Stock         5 non-null      int64  
dtypes: float64(1), int64(6), object(2)
memory usage: 400.0+ bytes
None
   Unnamed: 0_x  SaleID  Product_ID  Quantity  Price    SaleDate  \
0             0       1         101         2  19.99  2024-11-01   
1             1       2         102         1  29.99  2024-11-02   
2             2       3         103         3  27.49  2024-11-03   
3             3       4         101         5  19.99  2024-11-04

## Task 4: Group and Aggregate Data
1. Summarise the sales data by grouping it based on the product names and calculating the total quantities sold and total prices for each product.

    - Identify the column that will act as the group.
    - Decide on the aggregation functions to apply for other columns.

In [38]:

# Task 4: Group and Aggregate Data
# Group and aggregate data
# TODO: Group by products names and calculate total quantities sold and prices for each product.
grouped_data = merged_data.groupby('ProductName').agg({'Quantity': 'sum', 'Price': 'sum'}) 
print("\nGrouped Data (Total Sales per Product):")
print(grouped_data)
    


Grouped Data (Total Sales per Product):
               Quantity  Price
ProductName                   
Doohickey D           2  39.99
Gadget B              1  29.99
Thingamajig C         3  27.49
Widget A              7  39.98


## Task 5: Pivot Table

1. Create a table that summarises the quantity of products sold on each date, with product names as rows and dates as columns.

    - Think about which columns will serve as the `rows` (index) and columns in your table.
    - Identify the `values` you want to summarise and how they should be aggregated (e.g., summed up).

In [45]:
# Task 5: Pivot Table
# Create a pivot table
# TODO: Create a pivot table to summarise the quantity of products sold on each date, with product names as rows and dates as columns.
pivot_table = merged_data.pivot_table(index='ProductName', columns='SaleDate', values='Quantity', aggfunc='sum')  # Define index, columns, values, and aggfunc
# Replace NaN values with empty strings
pivot_table = pivot_table.fillna('')
print("\nPivot Table (Sales by Product and Date):")
print(pivot_table)


Pivot Table (Sales by Product and Date):
SaleDate      2024-11-01 2024-11-02 2024-11-03 2024-11-04 2024-11-05
ProductName                                                         
Doohickey D                                                      2.0
Gadget B                        1.0                                 
Thingamajig C                              3.0                      
Widget A             2.0                              5.0           


Your grouped data and pivot table are expected to look as follows:


### Grouped Data (Total Sales per Product)

| Product        | Quantity | Price  |
|----------------|----------|--------|
| Doohickey D    | 2        | 39.99  |
| Gadget B       | 1        | 29.99  |
| Thingamajig C  | 3        | 27.49  |
| Widget A       | 7        | 39.98  |

### Pivot Table (Sales by Product and Date)

| Product        | 2024-11-01 | 2024-11-02 | 2024-11-03 | 2024-11-04 | 2024-11-05 |
|----------------|------------|------------|------------|------------|------------|
| Doohickey D    |            |            |            |            | 2.0        |
| Gadget B       |            | 1.0        |            |            |            |
| Thingamajig C  |            |            | 3.0        |            |            |
| Widget A       | 2.0        |            |            | 5.0        |            |